models/parser.py

```diff
@@ -0,0 +1,232 @@
+        """
+        Implements the and eval function.
+        This is the intersection between the result of eval function of the two children nodes.
+        """
+        return self.left_expression.eval(index).intersection(self.right_expression.eval(index))
+
+    def __str__(self):
+        return '({} && {})'.format(self.left_expression, self.right_expression)
+
+
+class OrNode(BinaryNode):
+    """This class defines the OR operator logic in our tree."""
+
+    def eval(self, index):
+        """
+        Implements the or eval function.
+        This is the union between the result of eval function of the two children nodes.
+        """
+        return self.left_expression.eval(index).union(self.right_expression.eval(index))
+
+    def __str__(self):
+        return '({} || {})'.format(self.left_expression, self.right_expression)
+
+
+class NotNode(UnaryNode):
+    """This class defines the NOT operator logic in our tree."""
+
+    def eval(self, index):
+        """
+        Implements the not eval function.
+        This is the difference between:
+            - all posting_ids in index
+            - the result of eval function of the child node.
+        """
+        all_posting_ids = {
+            posting_id
+            for _, posting_list in index.values()
+            for posting_id, _, _ in posting_list
+        }
+        return all_posting_ids.difference(self.expression.eval(index))
+
+    def __str__(self):
+        return '!({})'.format(self.expression)
+
+
+class RoleNode(Node):
+    """This class defines the logic for leaves in our tree."""
+
+    def __init__(self, role_name):
+        """self.role_name is the value of the leaf (a word)"""
+        self.role_name = role_name
+
+    def eval(self, index):
+        """
+        Implements the eval function for RoleNode.
+        Return all posting_ids associated to self.role_name.
+        """
+        return {posting_id for posting_id, _, _, in index[self.role_name][1]}
+
+    def __str__(self):
+        return str(self.role_name)
+
+
+class Tokenizer(object):
+    """This class defines all the logic to tokenize a boolean expression."""
+
+    @staticmethod
+    def tokenize(text):
+        """
+        This methods tokenizes the given text input.
+        + params:
+            - text: string to tokenize
+        + returns:
+            a list of tokens
+        """
+        tokenized_text = []
+        current_token = ''
+
+        for char in text:
+            if char == ' ':
+                continue
+
+            if char in '!()':
+                if len(current_token) > 0:
+                    tokenized_text.append(current_token)
+                    current_token = ''
+                tokenized_text.append(char)
+                continue
+
+            if char in '|&':
+                if len(current_token) > 0:
+                    if char == current_token[-1]:
+                        tokenized_text.append(current_token[:-1]) if current_token[:-1] else None
```

Comment:
Is `else None` usefull here ? We just want to add the token to the list if it exists, say we do nothing otherwise seems useless

Reply:
Yes but it is not possible in python to write an if/else inline statement without the else part :'( Maybe there is another way of writing it properly but I don't see it.

Reply:
I did not knew that, ok for me